### ISSUE TITLE
Undefined function errors in heex file

### ISSUE THREAD

**Issue by bangalcat** [lexical-lsp/lexical#495]
Hi, thank you for all of your efforts on this awesome lsp.

I'm not sure that lexical fully supports heex files, though I'd like to report this for the records.

When I use function components by import, lexical diagnoses it as compile error:

<img width="837" alt="image" src="https://github.com/lexical-lsp/lexical/assets/16166083/3109c372-1f90-48d2-b176-0cd7a2217062">

If I use it with the module name, it doesn't emit error messages:

<img width="831" alt="image" src="https://github.com/lexical-lsp/lexical/assets/16166083/9abbc08c-ddcb-4740-98cc-8560659e4035">

I guess this occurs all functions without module name in heex file:

<img width="1041" alt="image" src="https://github.com/lexical-lsp/lexical/assets/16166083/afebb502-0cbf-4b28-903f-4d8ccaf6e6e4">

But mix compile works well anyway.

### My Environments

- lexical build: elixir 1.15.7 / otp 26.1.2 / on latest commit
- project : elixir 1.15.7 / otp 26.1.2
- MacOS Sonoma, using neovim, asdf

**Comment by scottming** [comment]
Hi, @bangalcat Heex is not fully supported, and this behavior is expected. After saving, this error should disappear. Currently, it is difficult to determine which `ex/heex` files are referenced by a `heex` file while you are editing it. We will further optimize the compilation diagnostics feature for `heex` once indexing is completed.

**Comment by bangalcat** [comment]
@scottming thanks for the quick response.

> After saving, this error should disappear. 

Hmm, as for me, the errors didn't disappear after saving. When I just open the file, the errors don't show, but when I start typing or saving the file, the errors appear. Am I missing something?

**Comment by scottming** [comment]
What happens when you run `mix compile` in the terminal? Essentially, when you save a file in Lexical, it parses the diagnostics generated by `mix compile`.

**Comment by bangalcat** [comment]
It passes without any error or warning messages. I've also tried `mix do clean, compile` but it's still the same.

**Comment by scottming** [comment]
@bangalcat Did you find any error in the `.lexical/lexical.log` or `.lexical/project.log`?

> as for me, the errors didn't disappear after saving.
since I can't replicate this: 

![CleanShot 2023-11-19 at 19 23 25@2x](https://github.com/lexical-lsp/lexical/assets/12830256/b993e067-4008-497b-b8ed-b0437bd198a4)

**Comment by bangalcat** [comment]
@scottming Oh, I just noticed an error message that I hadn't seen before.

```
2023-11-19T20:49:00.213971+09:00 error: :gen_event handler LXical.RemoteControl.Dispatch.Handlers.Indexing installed in LXical.RemoteControl.Dispatch terminating, ** (stop) exited in: GenServer.call(LXical.RemoteControl.Search.Store, {:update, "/(my path)/pento/lib/pento_web/components/layouts/root.html.heex", []}, 5000), ** (EXIT) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started, Last message: {:file_compile_requested, nil, 0, "file:///(my path)/pento/lib/pento_web/components/layouts/root.html.heex"}, State: []
```

And I just realized that I had enabled the indexing just for the trial and I'd forgotten about it 😅. After turning off the indexing it is working well!

**Comment by bangalcat** [comment]
Bad news. the error message appears again. I think at first it didn't emit the error due to the whole recompilation process. The error message shown in project.log file is slightly different now. 


https://github.com/lexical-lsp/lexical/assets/16166083/46df6182-7157-40bd-b9df-0b26294a364c



```
2023-11-19T21:54:45.035104+09:00 error: Generic event handler 'Elixir.LXical.RemoteControl.Dispatch.Handlers.Indexing' crashed. Installed: 'Elixir.LXical.RemoteControl.Dispatch'. Last event: {file_compile_requested,nil,0,<<"file:///(my path)/pento/lib/pento_web/live/faq_live/index.html.heex">>}. State: []. Reason: {noproc,{'Elixir.GenServer',call,['Elixir.LXical.RemoteControl.Search.Store',{update,<<"(my path)/pento/lib/pento_web/live/faq_live/index.html.heex">>,[]},5000]}}.

```

I'm not sure it is related. I also tested it using lexical build with elixir 1.14/erlang 26.1.2. I thought it was related to indexing but now I have no idea🤷‍♂️.

**Comment by scottming** [comment]
@bangalcat 

The issue seems to be caused by the indexing remaining `enabled`. Could you provide more information about the indexing error? As the indexing feature is still in its early stages, there may be some bugs.

If you don't see any more errors in `project.log`, could you try using version `1.14.5-otp-25`?

Additionally, if you want to temporarily resolve this issue, I suspect there might be a caching problem with our indexing feature. Even if you use `mix package` to build, this line: https://github.com/lexical-lsp/lexical/blob/c2113daaf48d9ad92086a4973584497eec848788/apps/common/lib/lexical/features.ex#L5 might still return true. To address this, you can modify the file by appending `dbg` and then re-run `mix package`. This should help resolve the issue.

cc @scohen

**Comment by bangalcat** [comment]
Alright. After I've been digging a while, although I don't want to rush to conclusions, I doubt it's an indexing issue. 
First, I followed your directions and it seems the above code obviously works correctly.  
Also, if I re-enable the indexing flag, there is no error messages both `project.log` and `lexical.log` files. 
Finally, for the test, I've cleared all of my configs and started with your [minimal configuration](https://github.com/scottming/nvim-mini-for-lexical/blob/master/init.lua) and `kickstart.nvim`. Then it works the same as your video. So I guess it could be related to my neovim configs. Btw, that's all my digging so far. I will track this more later, but for now this seems not to be an urgent issue.

**Comment by scohen** [comment]
indexing being abled is a compiled-in feature. If you need to turn it off, you need to recompile the project.

**Comment by scohen** [comment]
@bangalcat There should be more errors above the one you pasted. Can you find them?

**Comment by bangalcat** [comment]
@scohen No, I can't so far. There is no error except `error: Generic event handler 'Elixir.LXical.RemoteControl.Dispatch.Handlers.Indexing' crashed..` (when I disable the indexing). I will report back if I see something else.

**Comment by bangalcat** [comment]
Today I investigated this issue further using my free time. Here's what I've figured out so far.

Tl;dr: I suspect that the whole project compilation is done before the current file compilation in lexical server, at least in my environment. 

Every time I save or format the current file, the logs appear in order like this(project.log):

```
info: Compile completed with status ok Produced 0 diagnostics []
info: Lexical.RemoteControl.Build.State.compile_project/2 run and 0 diagnostics
info: Plugins found 0 results in 0.019 ms
info: Lexical.RemoteControl.Build.State.compile_file/2 run and 1 diagnostics # -> (undefined function error)
info: Plugins found 0 results in 0.006 ms
```

Btw, I found out that the file diagnostics result are coming from [here](https://github.com/lexical-lsp/lexical/blob/7ee08bce06d10ddbf5b2a7850c1a24695e6e1979/apps/remote_control/lib/lexical/remote_control/build/document/compilers/eex.ex#L49) `EEx.eval_quoted/2` which makes sense. 

So the errors should disappear after the whole project compilation, but under my neovim config the compile order is reversed. Maybe because the compile request order is reversed. I don't know why yet, but I will look into that soon.

**Comment by bangalcat** [comment]
Okay, I totally get it now. I'm using LazyVim for my neovim configuration, and it enables the 'autoformat' option in neovim as default. So whenever I save the file, a 'format' request sends to lexical. And it seems the 'format' request triggers 'compile file' in lexical server. 

I just turned off 'autoformat' option only for heex files, and it works like a charm. Although I still wish there would be full lsp supports of heex files, I'm happy with it now. This issue can be closed now.
